<a href="https://colab.research.google.com/github/Kauanldsbarbosa/predictive_maintenance_system/blob/main/predictive_maintenance_system.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Predictive Maintenance System



**About:**
As a way to consolidate my machine learning studies, this project aims to apply my knowledge so far by solving the following problem: Predicting whether an industrial machine will fail based on data from multiple sensors (temperature, vibration, pressure, etc.).


datasheet reference:


AI4I 2020 Predictive Maintenance Dataset [Dataset]. (2020). UCI Machine Learning Repository. https://doi.org/10.24432/C5HS5C.

## Import Dataset

In [2]:
!wget https://archive.ics.uci.edu/static/public/601/ai4i+2020+predictive+maintenance+dataset.zip -O dataset.zip && unzip dataset.zip && rm dataset.zip

--2026-04-25 18:22:53--  https://archive.ics.uci.edu/static/public/601/ai4i+2020+predictive+maintenance+dataset.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified
Saving to: ‘dataset.zip’

dataset.zip             [  <=>               ] 509.93K  2.06MB/s    in 0.2s    

2026-04-25 18:22:53 (2.06 MB/s) - ‘dataset.zip’ saved [522170]

Archive:  dataset.zip
replace ai4i2020.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

In [2]:
import pandas as pd

df = pd.read_csv('/content/ai4i2020.csv')
df.head()

,UDI,Product ID,Type,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Machine failure,TWF,HDF,PWF,OSF,RNF
0,1,M14860,M,298.1,308.6,1551,42.8,0,0,0,0,0,0,0
1,2,L47181,L,298.2,308.7,1408,46.3,3,0,0,0,0,0,0
2,3,L47182,L,298.1,308.5,1498,49.4,5,0,0,0,0,0,0
3,4,L47183,L,298.2,308.6,1433,39.5,7,0,0,0,0,0,0
4,5,L47184,L,298.2,308.7,1408,40.0,9,0,0,0,0,0,0


###Exploring datasheet

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   UDI                      10000 non-null  int64  
 1   Product ID               10000 non-null  object 
 2   Type                     10000 non-null  object 
 3   Air temperature [K]      10000 non-null  float64
 4   Process temperature [K]  10000 non-null  float64
 5   Rotational speed [rpm]   10000 non-null  int64  
 6   Torque [Nm]              10000 non-null  float64
 7   Tool wear [min]          10000 non-null  int64  
 8   Machine failure          10000 non-null  int64  
 9   TWF                      10000 non-null  int64  
 10  HDF                      10000 non-null  int64  
 11  PWF                      10000 non-null  int64  
 12  OSF                      10000 non-null  int64  
 13  RNF                      10000 non-null  int64  
dtypes: float64(3), int64(9)

In [4]:
df.describe()

,UDI,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Machine failure,TWF,HDF,PWF,OSF,RNF
count,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000
mean,5000.50000,300.004930,310.005560,1538.776100,39.986910,107.951000,0.033900,0.004600,0.011500,0.009500,0.009800,0.00190
std,2886.89568,2.000259,1.483734,179.284096,9.968934,63.654147,0.180981,0.067671,0.106625,0.097009,0.098514,0.04355
min,1.00000,295.300000,305.700000,1168.000000,3.800000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
25%,2500.75000,298.300000,308.800000,1423.000000,33.200000,53.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
50%,5000.50000,300.100000,310.100000,1503.000000,40.100000,108.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
75%,7500.25000,301.500000,311.100000,1612.000000,46.800000,162.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
max,10000.00000,304.500000,313.800000,2886.000000,76.600000,253.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.00000


In [5]:
df['Machine failure'].value_counts(), df['TWF'].value_counts(), df['HDF'].value_counts(), df['PWF'].value_counts(), df['OSF'].value_counts(), df['RNF'].value_counts()

(Machine failure
 0    9661
 1     339
 Name: count, dtype: int64,
 TWF
 0    9954
 1      46
 Name: count, dtype: int64,
 HDF
 0    9885
 1     115
 Name: count, dtype: int64,
 PWF
 0    9905
 1      95
 Name: count, dtype: int64,
 OSF
 0    9902
 1      98
 Name: count, dtype: int64,
 RNF
 0    9981
 1      19
 Name: count, dtype: int64)

In [6]:
corr_matrix = df.corr(numeric_only=True)
corr_matrix['Machine failure'].sort_values(ascending=False)

,Machine failure
Machine failure,1.000000
HDF,0.575800
OSF,0.531083
PWF,0.522812
TWF,0.362904
Torque [Nm],0.191321
Tool wear [min],0.105448
Air temperature [K],0.082556
Process temperature [K],0.035946
RNF,0.004516


In [7]:
X, y = df.drop(['UDI', 'Product ID', 'Machine failure', 'TWF', 'HDF', 'PWF', 'OSF', 'RNF'], axis=1), df['Machine failure']

In [8]:
X['Type'].value_counts()

,count
Type,
L,6000
M,2997
H,1003


In [9]:
X.shape, y.shape

((10000, 6), (10000,))

In [10]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

In [11]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((8000, 6), (2000, 6), (8000,), (2000,))

In [12]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8000 entries, 4058 to 1424
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Type                     8000 non-null   object 
 1   Air temperature [K]      8000 non-null   float64
 2   Process temperature [K]  8000 non-null   float64
 3   Rotational speed [rpm]   8000 non-null   int64  
 4   Torque [Nm]              8000 non-null   float64
 5   Tool wear [min]          8000 non-null   int64  
dtypes: float64(3), int64(2), object(1)
memory usage: 437.5+ KB


In [13]:
num_attribs = ['Air temperature [K]', 'Process temperature [K]', 'Rotational speed [rpm]', 'Torque [Nm]', 'Tool wear [min]']
cat_attribs = ['Type']

##Creante Pipeline



In [14]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), num_attribs),
        ('cat', OneHotEncoder(), cat_attribs)
    ]
)

X_prepared = preprocessor.fit_transform(X_train)

##Training Models

In [113]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import RandomizedSearchCV
from imblearn.over_sampling import SMOTE

###SVC

In [109]:
from sklearn.svm import SVC
from scipy.stats import loguniform
from imblearn.pipeline import Pipeline as ImbPipeline

svc_pipeline = ImbPipeline([
    ('preprocessor', preprocessor),
    ('smote', SMOTE(random_state=42)),
    ('svc', SVC(class_weight='balanced', probability=True))
])

params = {
    'svc__C': loguniform(1e-3, 1e2),
    'svc__kernel': ['poly', 'rbf', 'sigmoid'],
    'svc__gamma': loguniform(1e-4, 1e1),
    'svc__degree': [2, 3, 4]
}

svc_random_search = RandomizedSearchCV(
    estimator=svc_pipeline,
    param_distributions=params,
    n_iter=5,
    cv=5,
    random_state=42,
    n_jobs=-1,
    verbose=2,
    scoring='recall'
    )

svc_random_search.fit(X_train, y_train)

Fitting 5 folds for each of 5 candidates, totalling 25 fits


RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('preprocessor',
                                              ColumnTransformer(transformers=[('num',
                                                                               StandardScaler(),
                                                                               ['Air '
                                                                                'temperature '
                                                                                '[K]',
                                                                                'Process '
                                                                                'temperature '
                                                                                '[K]',
                                                                                'Rotational '
                                                                                'speed '
                                                                                '[rpm]',
                                                                                'Torque '
                                                                                '[Nm]',
                                                                                'Tool '
                                                                                'wear '
                                                                                '[min]']),
                                                                              ('cat',
                                                                               OneHotEncoder(),
                                                                               ['Type'])])),
                                             ('smote', SMOTE(random_state=42)),
                                             ('svc',
                                              SVC(class_weight='balanced',
                                                  probability=True))]),
                   n_iter=5, n_jobs=-1,
                   param_distributions={'svc__C': <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7fcc1c86a1e0>,
                                        'svc__degree': [2, 3, 4],
                                        'svc__gamma': <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7fcc1c86a900>,
                                        'svc__kernel': ['poly', 'rbf',
                                                        'sigmoid']},
                   random_state=42, scoring='recall', verbose=2)

In [89]:
svc_random_search.best_params_, svc_random_search.best_score_

({'C': np.float64(0.011526449540315618),
  'degree': 2,
  'gamma': np.float64(0.122294675253906),
  'kernel': 'rbf'},
 np.float64(0.9083964045433042))

####Evaluating

In [95]:
from sklearn.metrics import classification_report

svc_best_model = svc_random_search.best_estimator_
X_test_prepared = preprocessor.transform(X_test)
y_pred = svc_best_model.predict(X_test_prepared)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.99      0.87      0.93      1932
           1       0.19      0.87      0.31        68

    accuracy                           0.87      2000
   macro avg       0.59      0.87      0.62      2000
weighted avg       0.97      0.87      0.91      2000



In [96]:
from sklearn.metrics import confusion_matrix
from matplotlib import pyplot as plt

confusion_matrix(y_test, y_pred)


array([[1683,  249],
       [   9,   59]])

###SGDClassifier

In [114]:
from sklearn.linear_model import SGDClassifier
from imblearn.pipeline import Pipeline as ImbPipeline

sgdc_pipeline = ImbPipeline([
    ('preprocessor', preprocessor),
    ('smote', SMOTE(random_state=42)),
    ('sgdc', SGDClassifier(loss='log_loss'))
])

params = {
    'sgdc__loss': ['log_loss', 'modified_huber'],
    'sgdc__penalty': ['l2', 'l1', 'elasticnet'],
    'sgdc__alpha': [1e-4, 1e-3, 1e-2, 1e-1],
    'sgdc__learning_rate': ['optimal', 'adaptive'],
    'sgdc__eta0': [0.01, 0.1]
}
sgdc_random_search = RandomizedSearchCV(
    estimator=sgdc_pipeline,
    param_distributions=params,
    n_iter=5,
    cv=5,
    random_state=42,
    n_jobs=-1,
    verbose=2,
    scoring='recall'
    )

sgdc_random_search.fit(X_train, y_train)

Fitting 5 folds for each of 5 candidates, totalling 25 fits


RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('preprocessor',
                                              ColumnTransformer(transformers=[('num',
                                                                               StandardScaler(),
                                                                               ['Air '
                                                                                'temperature '
                                                                                '[K]',
                                                                                'Process '
                                                                                'temperature '
                                                                                '[K]',
                                                                                'Rotational '
                                                                                'speed '
                                                                                '[rpm]',
                                                                                'Torque '
                                                                                '[Nm]',
                                                                                'Tool '
                                                                                'wear '
                                                                                '[min]']),
                                                                              ('cat',
                                                                               OneHotEncoder(),
                                                                               ['Type'])])),
                                             ('smote', SMOTE(random_state=42)),
                                             ('sgdc',
                                              SGDClassifier(loss='log_loss'))]),
                   n_iter=5, n_jobs=-1,
                   param_distributions={'sgdc__alpha': [0.0001, 0.001, 0.01,
                                                        0.1],
                                        'sgdc__eta0': [0.01, 0.1],
                                        'sgdc__learning_rate': ['optimal',
                                                                'adaptive'],
                                        'sgdc__loss': ['log_loss',
                                                       'modified_huber'],
                                        'sgdc__penalty': ['l2', 'l1',
                                                          'elasticnet']},
                   random_state=42, scoring='recall', verbose=2)

####Evaluating

In [115]:
sgdc_random_search.best_params_, sgdc_random_search.best_score_

({'sgdc__penalty': 'l2',
  'sgdc__loss': 'modified_huber',
  'sgdc__learning_rate': 'adaptive',
  'sgdc__eta0': 0.01,
  'sgdc__alpha': 0.001},
 np.float64(0.8010774410774409))

In [116]:
from sklearn.metrics import classification_report

sgdc_best_model = sgdc_random_search.best_estimator_
y_pred = sgdc_best_model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.99      0.83      0.90      1932
           1       0.14      0.81      0.24        68

    accuracy                           0.83      2000
   macro avg       0.57      0.82      0.57      2000
weighted avg       0.96      0.83      0.88      2000



In [117]:
from sklearn.metrics import confusion_matrix
from matplotlib import pyplot as plt

confusion_matrix(y_test, y_pred)

array([[1604,  328],
       [  13,   55]])

###Random Florest

In [103]:
from sklearn.ensemble import RandomForestClassifier
from imblearn.pipeline import Pipeline as ImbPipeline

rnd_florest_pipeline = ImbPipeline([
    ('preprocessor', preprocessor),
    ('smote', SMOTE(random_state=42)),
    ('rnd_florest', RandomForestClassifier())
])

params = {
    'rnd_florest__n_estimators': [100, 200, 500],
    'rnd_florest__max_depth': [None, 10, 20, 30],
    'rnd_florest__min_samples_split': [2, 5, 10],
    'rnd_florest__min_samples_leaf': [1, 2, 4],
    'rnd_florest__max_features': ['sqrt', 'log2', None],
    'rnd_florest__bootstrap': [True, False]
}

rnd_florest_pipeline_random_search = RandomizedSearchCV(
    estimator=rnd_florest_pipeline,
    param_distributions=params,
    n_iter=5,
    cv=5,
    random_state=42,
    n_jobs=-1,
    verbose=2,
    scoring='recall'
    )

rnd_florest_pipeline_random_search.fit(X_train, y_train)

Fitting 5 folds for each of 5 candidates, totalling 25 fits


RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('preprocessor',
                                              ColumnTransformer(transformers=[('num',
                                                                               StandardScaler(),
                                                                               ['Air '
                                                                                'temperature '
                                                                                '[K]',
                                                                                'Process '
                                                                                'temperature '
                                                                                '[K]',
                                                                                'Rotational '
                                                                                'speed '
                                                                                '[rpm]',
                                                                                'Torque '
                                                                                '[Nm]',
                                                                                'Tool '
                                                                                'wear '
                                                                                '[min]']),
                                                                              ('cat',
                                                                               OneHotEncoder(),
                                                                               ['Type'])])),
                                             ('smote', SMOTE(random_state=42)),
                                             ('rnd_florest',
                                              RandomForestClassifier())]),
                   n_iter=5, n_jobs=-1,
                   param_distributions={'rnd_florest__bootstrap': [True, False],
                                        'rnd_florest__max_depth': [None, 10, 20,
                                                                   30],
                                        'rnd_florest__max_features': ['sqrt',
                                                                      'log2',
                                                                      None],
                                        'rnd_florest__min_samples_leaf': [1, 2,
                                                                          4],
                                        'rnd_florest__min_samples_split': [2, 5,
                                                                           10],
                                        'rnd_florest__n_estimators': [100, 200,
                                                                      500]},
                   random_state=42, scoring='recall', verbose=2)

####Evaluating

In [104]:
rnd_florest_pipeline_random_search.best_estimator_, rnd_florest_pipeline_random_search.best_score_

(Pipeline(steps=[('preprocessor',
                  ColumnTransformer(transformers=[('num', StandardScaler(),
                                                   ['Air temperature [K]',
                                                    'Process temperature [K]',
                                                    'Rotational speed [rpm]',
                                                    'Torque [Nm]',
                                                    'Tool wear [min]']),
                                                  ('cat', OneHotEncoder(),
                                                   ['Type'])])),
                 ('smote', SMOTE(random_state=42)),
                 ('rnd_florest',
                  RandomForestClassifier(max_depth=10, min_samples_leaf=4,
                                         min_samples_split=5))]),
 np.float64(0.8191245791245791))

In [105]:
from sklearn.metrics import classification_report

random_florest_best_model = rnd_florest_pipeline_random_search.best_estimator_
y_pred = random_florest_best_model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.99      0.95      0.97      1932
           1       0.38      0.79      0.51        68

    accuracy                           0.95      2000
   macro avg       0.69      0.87      0.74      2000
weighted avg       0.97      0.95      0.96      2000



In [106]:
from sklearn.metrics import confusion_matrix
from matplotlib import pyplot as plt

random_florest_best_model = rnd_florest_pipeline_random_search.best_estimator_
y_pred = random_florest_best_model.predict(X_test)
confusion_matrix(y_test, y_pred)

array([[1844,   88],
       [  14,   54]])

##Ensemble

###VotingClassifier

In [118]:
from sklearn.ensemble import VotingClassifier
from imblearn.pipeline import Pipeline as ImbPipeline

voting_clf = VotingClassifier(
    estimators=[
        ('svc', svc_random_search.best_estimator_),
        ('sgdc', sgdc_best_model),
        ('rnd_florest', rnd_florest_pipeline_random_search.best_estimator_)
    ],
    voting='soft'
)

voting_clf_pipeline = ImbPipeline([
    ('preprocessor', preprocessor),
    ('smote', SMOTE(random_state=42)),
    ('voting_clf', voting_clf)
])


voting_clf.fit(X_train, y_train)

VotingClassifier(estimators=[('svc',
                              Pipeline(steps=[('preprocessor',
                                               ColumnTransformer(transformers=[('num',
                                                                                StandardScaler(),
                                                                                ['Air '
                                                                                 'temperature '
                                                                                 '[K]',
                                                                                 'Process '
                                                                                 'temperature '
                                                                                 '[K]',
                                                                                 'Rotational '
                                                                                 'speed '
                                                                                 '[rpm]',
                                                                                 'Torque '
                                                                                 '[Nm]',
                                                                                 'Tool '
                                                                                 'wear '
                                                                                 '[min]']),
                                                                               ('cat',
                                                                                OneHotEncoder(),
                                                                                ['Type'])])),
                                              ('smote', SMOTE(random_state=42)),
                                              ('svc',
                                               SVC(C=np.float64(0.01152644954031...
                                               ColumnTransformer(transformers=[('num',
                                                                                StandardScaler(),
                                                                                ['Air '
                                                                                 'temperature '
                                                                                 '[K]',
                                                                                 'Process '
                                                                                 'temperature '
                                                                                 '[K]',
                                                                                 'Rotational '
                                                                                 'speed '
                                                                                 '[rpm]',
                                                                                 'Torque '
                                                                                 '[Nm]',
                                                                                 'Tool '
                                                                                 'wear '
                                                                                 '[min]']),
                                                                               ('cat',
                                                                                OneHotEncoder(),
                                                                                ['Type'])])),
                                              ('smote', SMOTE(random_state=42)),
                                              ('rnd_florest',
                                               RandomForestClassif

In [119]:
from sklearn.metrics import classification_report

y_pred = voting_clf.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.99      0.92      0.96      1932
           1       0.28      0.84      0.42        68

    accuracy                           0.92      2000
   macro avg       0.64      0.88      0.69      2000
weighted avg       0.97      0.92      0.94      2000



In [120]:
from sklearn.metrics import confusion_matrix
from matplotlib import pyplot as plt

confusion_matrix(y_test, y_pred)

array([[1786,  146],
       [  11,   57]])

In [123]:
from sklearn.metrics import accuracy_score, recall_score

print(accuracy_score(y_test, y_pred))
print(recall_score(y_test, y_pred))

0.9215
0.8382352941176471


###Stacking

In [124]:
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import LogisticRegression
from imblearn.pipeline import Pipeline as ImbPipeline

stacking_clf = StackingClassifier(
    estimators=[
        ('svc', svc_random_search.best_estimator_),
        ('sgdc', sgdc_best_model),
        ('rnd_florest', rnd_florest_pipeline_random_search.best_estimator_)
    ],
    final_estimator=LogisticRegression(random_state=42),
    n_jobs=-1
)

stacking_clf_pipeline = ImbPipeline([
    ('preprocessor', preprocessor),
    ('smote', SMOTE(random_state=42)),
    ('stacking_clf', stacking_clf)
])


stacking_clf.fit(X_train, y_train)

StackingClassifier(estimators=[('svc',
                                Pipeline(steps=[('preprocessor',
                                                 ColumnTransformer(transformers=[('num',
                                                                                  StandardScaler(),
                                                                                  ['Air '
                                                                                   'temperature '
                                                                                   '[K]',
                                                                                   'Process '
                                                                                   'temperature '
                                                                                   '[K]',
                                                                                   'Rotational '
                                                                                   'speed '
                                                                                   '[rpm]',
                                                                                   'Torque '
                                                                                   '[Nm]',
                                                                                   'Tool '
                                                                                   'wear '
                                                                                   '[min]']),
                                                                                 ('cat',
                                                                                  OneHotEncoder(),
                                                                                  ['Type'])])),
                                                ('smote',
                                                 SMOTE(random_state=42)),
                                                ('svc',
                                                 SVC(C=np.float64(0.011526449540...
                                                                                  StandardScaler(),
                                                                                  ['Air '
                                                                                   'temperature '
                                                                                   '[K]',
                                                                                   'Process '
                                                                                   'temperature '
                                                                                   '[K]',
                                                                                   'Rotational '
                                                                                   'speed '
                                                                                   '[rpm]',
                                                                                   'Torque '
                                                                                   '[Nm]',
                                                                                   'Tool '
                                                                                   'wear '
                                                                                   '[min]']),
                                                                                 ('cat',
                                                                                  OneHotEncoder(),
                                                                                  ['Type'])])),
                                                ('smote',
                                                 SMOTE(random_state=42)),
                                 

In [125]:
from sklearn.metrics import classification_report

y_pred = stacking_clf.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99      1932
           1       0.71      0.62      0.66        68

    accuracy                           0.98      2000
   macro avg       0.85      0.80      0.83      2000
weighted avg       0.98      0.98      0.98      2000



In [126]:
from sklearn.metrics import confusion_matrix
from matplotlib import pyplot as plt

confusion_matrix(y_test, y_pred)

array([[1915,   17],
       [  26,   42]])

In [127]:
from sklearn.metrics import accuracy_score, recall_score

print(accuracy_score(y_test, y_pred))
print(recall_score(y_test, y_pred))

0.9785
0.6176470588235294


##Conclusion and Choice of the Final Model

After evaluating different ensemble strategies to handle the unbalanced nature of the data (previously processed with SMOTE), the Voting Classifier (Soft Voting) was chosen as the definitive model for this problem.

During testing, a Stacking architecture (using Logistic Regression as a meta-model) was also implemented. Although Stacking increased Precision to almost 98%, it exhibited excessively conservative behavior, shifting the decision boundary and reducing the recall rate of true positives to approximately 61%.

Considering the context of the problem, where detecting the minority class is crucial and failing to identify it (false negatives) has a high cost, this drop in recall makes the use of Stacking unfeasible. The Voting Classifier, on the other hand, delivered the best balance, maintaining excellent Precision (~92%) while simultaneously achieving a robust recall (~83%), ensuring that the vast majority of real events are detected with a low rate of false alarms.